File: DataStructures/DLL/reverse.java
class LinkedList
{
    static Node head;

    static class Node {
        int data;
        Node next, prev;

        Node(int d)
        {
            data = d;
            next = prev = null;
        }
    }

    void reverse()
    {
        Node temp = null;
        Node current = head;
        while(current != null)
        {
            temp = current.prev;
            current.prev = current.next;
            current.next = temp;
            current = current.prev;
        }
        if(temp != null && temp.next != null)
        {
            head = temp.prev;
        }
    }

    void push(int new_data)
    {
        Node new_node = new Node(new_data);
        new_node.prev = null;
        new_node.next = head;
        //this makes sense and has to be done now
        if(head != null)
        {
            head.prev = new_node;
        }

        head = new_node;
    }

    void printList(Node node)
    {
        while(node != null)
        {
            System.out.print(node.data + " ");
            node = node.next;
        }
    }

    public static void main(String[] args)
    {
        LinkedList list = new LinkedList();

        list.push(2);
        list.push(10);
        list.push(12);
        list.push(8);
        list.push(4);

        System.out.print("Original linked list");
        list.printList(head); 
        list.reverse();
        System.out.print("Reversed List");
        list.printList(head);
    }
}

File: DataStructures/DLL/reverse.c
#include <stdlib.h>
#include <stdio.h>

struct Node 
{
    int data;
    struct Node * prev;
    struct Node * next;
}

void push(struct Node ** head_ref, int new_data)
{
    struct Node * new_node = (struct Node * )malloc(sizeof(struct Node));
    new_node->data = new_data;
    new_node->prev = NULL;
    new_node->next = *head_ref;
    if(*head_ref != NULL)
    {
        (*head_ref)->prev = new_node;
    }
    (*head_ref) = new_node;
}

void reverse(struct Node ** head_ref)
{
    struct Node *temp = NULL;
    struct Node * current = (*head_ref);

    while(current)
    {
        temp = current->prev;
        current->prev = current->next;
        current->next = temp;
        current = current->prev;
    }

    if(temp != NULL)
    {
        *head_ref = temp->prev;
    }

}

void printList(struct Node * node)
{
    while(node != NULL)
    {
        printf("%d ", node->data);
        node = node-> next;
    }
}

int main()
{
    Node * head = NULL;
    push(&head, 36);
    push(&head, 35);
    push(&head, 34);
    push(&head, 33);
    push(&head, 32);

    printf("The original Linked List");
    printList(head);
    reverse(&head);
    printList("The reversed Linked List");
    printList(head);


    EXIT_SUCCESS(true);
    
}